scripts/util/doc-builder.js

```diff
@@ -0,0 +1,236 @@
+      if (docsBranchReference.isRemote()) {
+        console.info('info:', this.options.docsBranchName, 'is not available locally. But found in remote. Attempting to create');
+        return repo.getCommit(docsBranchReference.targetPeel() || docsBranchReference.target()).then(commit => {
+          return repo.createBranch(this.options.docsBranchName, commit, false, repo.defaultSignature(), 'Created docs branch in local repo');
+        }).then(reference => {
+          console.info('info:', 'Branch created successfully');
+          console.info('info:', 'Setting up', reference.shorthand(), 'to track', docsBranchReference.shorthand());
+          return NodeGit.Branch.setUpstream(reference, docsBranchReference.shorthand());
+        });
+      }
+    });
+  }).then(() => {
+    console.info('info:', this.options.docsBranchName, 'has been confirmed to exist');
+  });
+}
+
+DocBuilder.prototype.checkoutDocsBranch = function () {
+  if (fs.existsSync(this.options.docsDirName)) {
+    fsExtra.removeSync(this.options.docsDirName);
+  }
+
+  fs.mkdirSync(this.options.docsDirName);
+
+  console.info('\ninfo:', 'Checking out', this.options.docsBranchName, 'branch');
+  return NodeGit.Repository.open('.').then(repo => {
+
+    console.info('info:', 'Getting', this.options.docsBranchName, '\'s latest commit');
+    return repo.getBranchCommit(this.options.docsBranchName).then(commit => {
+      console.info('info:', 'Getting its working tree');
+      return commit.getTree();
+    }).then(tree => {
+      var options = {
+        checkoutStrategy: this.options.checkoutStrategy,
+        targetDirectory: path.join(repo.workdir(), this.options.docsDirName)
+      };
+
+      console.info('info:', 'Checking out the tree to', options.targetDirectory);
+      return NodeGit.Checkout.tree(repo, tree, options);
+    }).then(() => {
+      console.info('info:', 'Done checking out', this.options.docsBranchName);
+    });
+  });
+}
+
+DocBuilder.prototype.checkoutAPISrc = function () {
+  if (!fs.existsSync(this.options.apiDirPath)) {
+    fs.mkdirSync(this.options.apiDirPath);
+  }
+
+  console.info('\ninfo:', 'Checking out source code for each version tag and the', this.options.masterBranchName, 'branch');
+  return NodeGit.Repository.open('.').then(repo => {
+
+    console.info('info:', 'Getting available references within the repo');
+    return repo.getReferences(NodeGit.Reference.TYPE.OID).then(references => {
+      var promise = Promise.resolve();
+      var paths = [];
+
+      console.info('info:', 'Filtering them to include only tags and the', this.options.masterBranchName, 'branch.');
+      references.forEach(reference => {
+        var target = reference.targetPeel() || reference.target();
+        var name = reference.shorthand();
+        var isMasterBranch = reference.isBranch() && name === this.options.masterBranchName;
+
+        if (!isMasterBranch && !reference.isTag()) {
+          return;
+        }
+
+        promise = promise.then(() => {
```

Comment:
i've had times in which i "fry" my clone of the repo because of simultaneous changes to the repo. I'd like to think the "reads" can occur simultaneously, but i seriously doubt checking out asynchronously would be a good idea.

i'd explore and see how this can be better written.